RecoParticleFlow/PFProducer/src/MLPFModel.cc

```diff
@@ -0,0 +1,212 @@
+#include "RecoParticleFlow/PFProducer/interface/MLPFModel.h"
+
+namespace reco::mlpf {
+
+  //Prepares the input array of floats for a single PFElement
+  std::array<float, NUM_ELEMENT_FEATURES> get_element_properties(const reco::PFBlockElement& orig) {
+    const auto type = orig.type();
+    float pt = 0.0;
+    //these are placeholders for the the future
+    [[maybe_unused]] float deltap = 0.0;
+    [[maybe_unused]] float sigmadeltap = 0.0;
+    [[maybe_unused]] float px = 0.0;
+    [[maybe_unused]] float py = 0.0;
+    [[maybe_unused]] float pz = 0.0;
+    float eta = 0.0;
+    float phi = 0.0;
+    float energy = 0.0;
+    float trajpoint = 0.0;
+    float eta_ecal = 0.0;
+    float phi_ecal = 0.0;
+    float eta_hcal = 0.0;
+    float phi_hcal = 0.0;
+    float charge = 0;
+    float layer = 0;
+    float depth = 0;
+    float muon_dt_hits = 0.0;
+    float muon_csc_hits = 0.0;
+
+    if (type == reco::PFBlockElement::TRACK) {
+      const auto& matched_pftrack = orig.trackRefPF();
+      if (matched_pftrack.isNonnull()) {
+        const auto& atECAL = matched_pftrack->extrapolatedPoint(reco::PFTrajectoryPoint::ECALShowerMax);
+        const auto& atHCAL = matched_pftrack->extrapolatedPoint(reco::PFTrajectoryPoint::HCALEntrance);
+        if (atHCAL.isValid()) {
+          eta_hcal = atHCAL.positionREP().eta();
+          phi_hcal = atHCAL.positionREP().phi();
+        }
+        if (atECAL.isValid()) {
+          eta_ecal = atECAL.positionREP().eta();
+          phi_ecal = atECAL.positionREP().phi();
+        }
+      }
+      const auto& ref = ((const reco::PFBlockElementTrack*)&orig)->trackRef();
+      pt = ref->pt();
+      px = ref->px();
+      py = ref->py();
+      pz = ref->pz();
+      eta = ref->eta();
+      phi = ref->phi();
+      energy = ref->pt() * cosh(ref->eta());
```

Comment:
what's wrong with `p()`? it's faster and numerically more precise

Reply:
you're right.